doc/tutorial/intro-to-autoinstall.rst

```diff
@@ -0,0 +1,239 @@
+   Flash Drive, this can be done by copying the :code:`autoinstall.yaml` to the
+   partition containing the contents of the ISO - i.e., in the same directory
+   containing the :code:`casper` directory.
+ * On the rootfs of the installation system - this option will typically
+   require modifying the installation ISO and is not suggested, but is
+   supported.
+
+Directly specifying autoinstall as a :code:`autoinstall.yaml` file does not
+require a :code:`#cloud-config` header, and does not use a top level
+:code:`autoinstall` key.  The autoinstall directives are placed at the top
+level.
+
+.. code-block:: yaml
+
+    version: 1
+    ....
+
+
+Cloud-init and Autoinstall interaction
+======================================
+
+While cloud-init may assist in providing the autoinstall configuration to the
+Ubuntu installer, cloud-init itself is not processing the autoinstall.
+
+If cloud-init directives are intended to modify the ephemeral system, they
+must appear at the top level of the cloud-config.  If instead
+cloud-init directives are intended to modify the system being installed, they
+must appear under a :code:`user-data` section in :code:`autoinstall`.
+
+.. code-block:: yaml
+
+    #cloud-config
+    # cloud-init directives may optionally be specified here.
+    # These directives affect the ephemeral system performing the install.
+
+    autoinstall:
+      # autoinstall directives must be specified here, not directly at the top
+      # level.  These directives are processed by the Ubuntu Installer, and
+      # configure the target system to be installed.
+
+      user-data:
+        # cloud-init directives may also be optionally be specified here.
+        # These directives also affect the target system to be installed,
+        # and are processed on first boot.
+
+
+Zero-touch deployment with autoinstall
+======================================
+
+The Ubuntu Installer contains a safeguard, intended to prevent USB Flash Drives
+with an :code:`autoinstall.yaml` file from wiping out the wrong system.
+
+During autoinstall, a prompt will be shown to confirm that the install really
+should proceed and start making modifications to the target system. ::
+
+    start: subiquity/Meta/status_GET
+    Confirmation is required to continue.
+    Add 'autoinstall' to your kernel command line to avoid this
+
+
+    Continue with autoinstall? (yes|no)
+
+To bypass this prompt, arrange for the argument :code:`autoinstall` to be
+present on the kernel command line.
+
+
+Creating an autoinstall config
+==============================
+
+When any system is installed using the Ubuntu installer, an autoinstall file
+for repeating the install is created at
+:code:`/var/log/installer/autoinstall-user-data`.
+
+
+The structure of an autoinstall configuration
+=============================================
+
+See the :ref:`ai` for full details on the supported autoinstall directives.
+
+A minimal autoinstall configuration is:
+
+.. code-block:: yaml
+
```

Comment:
I'm a bit torn on providing examples without the `#cloud-config` + autoinstall key when we explicitly suggest to use the cloud-init approach. I know it's boilerplate so we probably don't want to repeat the same thing over and over in the examples.

Maybe a little reminder before or after the examples for people tempted to copy-paste? Something like:

> /!\ Remember to add the `#cloud-config` header and place the configuration within an `autoinstall` section if you are using the cloud-init approach.

```yaml
#cloud-config
autoinstall:
  < place the autoinstall configuration here >
```

Reply:
A goal I have for the doc overhaul is to reduce people sending autoinstall as cloud-config, so I think something like this would help reduce it.  I'm not sure I want the "Remember to add" part all the time, but to just have 99% of the autoinstall snippets (pretty much all of them except the case showing the non-cloud-init path in this document) just always include
```
#cloud-config
autoinstall:
```
might well help reduce that confusion.

Reply:
So yes, I like the suggestion very much, but think I want to skip the reminder part and just include the above 2 lines in yaml samples.

Reply:
Thanks!

Reply:
I think having a warning at the top where you introduce the #cloud-config and autoinstall lines would achieve the same thing (see my comment around line 64) - any user who wants to use that method will have their attention drawn to the fact that these lines are necessary - but having the lines also included in the examples is a great idea in case users try to copy/paste directly. A tutorial should be foolproof, so if we give them examples to try they shouldn't be able to fail with them if they follow the instructions.

Reply:
Yes, I'm erring on the side of foolproof.  We have adequate user evidence that there is a lot of user confusion on this point.  I'd normally prefer the shorter code sample.

Something I've debated doing is to allow direct specification of the `autoinstall.yaml` file to be in Cloud config data format - if `autoinstall` is the only top level directive, then this is simply handled subiquity side.  But what if there are other cloud-config top level keys?  I think Subiquity should warn or error in that case, as those aren't going to be processed.

If we did that, then 100% of the time it would be possible to just copy-paste the code sample in Cloud config data format and get a working result, regardless of if providing the data with the help of cloud-init or with modified install media.